fastcache.go

```diff
@@ -121,6 +122,26 @@ func (c *Cache) UpdateStats(s *Stats) {
 	}
 }
 
+// Keys retrieves all cached keys matching regex pattern
+func (c *Cache) Keys(pattern string) (keys [][]byte, err error) {
+	r, err := regexp.Compile(pattern)
+	if err != nil {
+		return
+	}
+
+	for _, b := range c.buckets {
+		for _, chunk := range b.chunks {
+			if len(chunk) > 0 {
+				if key := chunk[4 : 4+chunk[1]]; r.Match(key) {
+					keys = append(keys, key)
```

Comment:
`key` byte slice belongs to a `chunk`. It cannot be returned as-is to a caller without copying, since it may lead to data races.

Reply:
And if you make a copy, the memory usage could be high for a very large number of keys, so this leads to your first comment.

Reply:
Yes, that's why `VisitAllEntries` must be used instead. It has zero memory allocations irregardless of cache size. Memory allocations would depend only on the user-provided code inside `f` callback.